argoproj-labs/argocd-operator
# feat: Upgrade RH-SSO to v7.5.1 and support kube:admin, ocp groups and proxy env

**What type of PR is this?**
> /kind enhancement

**What does this PR do / why we need it**:

**This PR adds the below enhancements.**
-------------------------------------------
* Added missing RBAC annotations in `agocd_controller.go`: `template.openshift.io` and `oauth.openshift.io` api groups. Without these annotations, RBAC manifests won't get generated properly in order to install RHSSO.
* Removed hardcoded OIDC client secrets and replaced them by generated random strings. These client secrets do not need to be stored. 
* Added `KeycloakIdentityProviderMapper` to workaround https://github.com/keycloak/keycloak-operator/issues/471. Without the workaround, tokens can't be translated to credentials that contain the group identity.
* Propagated proxy environment variables from operator to Keycloak containers so that outbound calls can go through cluster proxy if cluste proxy is configured.
* Delete OAuth client using foreground propagation policy to ensure that the garbage collector removes all instantiated objects before the TemplateInstance itself disappears.
* Add extra Delete OAuth client calls to workaround https://github.com/openshift/client-go/issues/209
* Add retries to handle RHSSO image upgrade
* Upgrades the default version of RH-SSO to 7.5.1
* Adds support to login with `kube:admin` OpenShift User.
* With this PR, RH-SSO can fetch the group details of OpenShift Users(You should see the group details of the logged in user in the Argo CD console). This will help admins define Group Level RBAC for their OpenShift groups. 
   https://github.com/keycloak/keycloak/pull/8381
* Kube:admin by default you only have. Not working previously. not an OS user.
   https://github.com/keycloak/keycloak/pull/8428
* RHSSO works in proxy enabled cluster with no additional or manual configuration.
   https://github.com/keycloak/keycloak/pull/8559
*. Upgrades Keycloak version to `15.0.2`. On non-OpenShif Kubernetes clusters, users will see a Keycloak version upgrade from `9.0.3` to `15.0.2`

**Upgrades**
----------------
Whether you are using the Operator on Kubernetes or OpenShift, Version upgrades mentioned above are smooth and does not require any manual intervention other than upgrading the operator to new version(0.3.0) once it's released.

**Have you updated the necessary documentation?**
------------------

**How to test changes / Special notes to the reviewer**:
-------------------

Kubernetes:
----------------------
1. Deploy the below catalog source
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: argocd-catalog
spec:
  sourceType: grpc
  image: quay.io/aveerama/argocd-operator-index@sha256:bb0db86d3e6e27fe9d9e6891027db62e3b15f2947d65b059de8c7aae3a582eda
  displayName: Argo CD Operators
  publisher: Argo CD Community
```

2. Run the make target to install the CRDs `make install`
3. kubectl create namespace argocd
4. kubectl create -n argocd -f deploy/operator_group.yaml
5. kubectl create -n argocd -f deploy/subscription.yaml
6. kubectl create -n argocd -f examples/argocd-keycloak-k8s.yaml

OpenShift:
---------------
1. Deploy the below catalog source into `openshift-operators` namespace.
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: argocd-catalog
spec:
  sourceType: grpc
  image: quay.io/aveerama/argocd-operator-index@sha256:bb0db86d3e6e27fe9d9e6891027db62e3b15f2947d65b059de8c7aae3a582eda
  displayName: Argo CD Operators
  publisher: Argo CD Community
```
2. Move to Operator Hub to install the Operator.
3. kubectl create namespace argocd
4. kubectl create -n argocd -f examples/argocd-keycloak-openshift.yaml


Testing upgrade:
-------------------

1. Install the `0.2.0` of operator(same for k8s and OpenShift).
2. Create an Argo CD Instance with Keycloak as shown above.
3. Delete the operator but keep the Argo CD Instance and workloads as is.
4. Now, Install the new version of operator and see if the Keycloak pod is recreated and updated.

How to test login with kube:admin
------------------------

On your OpenShift cluster, Go to Networking -> Routes -> Click on the Argo CD route.
1. Click on `Login Via Keycloak` -> `Login with OpenShift` -> Provide `kubeadmin` credentials.
2. Once you login to Argo CD, you can also confirm this by looking into the user profile section.

Run E2E test:
--------------------
Download kuttl on your laptop or server
Run operator locally or through the bundle.
Run the below command to execute rhsso e2e tests
`kubectl kuttl test --config kuttl-test-rhsso.yaml`
